MarsAdvancedTaskPart1NUnitAutomationV4/Steps/LanguageSteps.cs

```diff
@@ -0,0 +1,190 @@
+using MarsAdvancedTaskPart1NUnitAutomation.Pages.ProfileOverview.ProfileNavigationMenuComponents.RenderingComponents;
+using MarsAdvancedTaskPart1NUnitAutomation.Utilities;
+using NUnit.Framework;
+using OpenQA.Selenium;
+using OpenQA.Selenium.Support.UI;
+using SeleniumExtras.WaitHelpers;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace MarsAdvancedTaskPart1NUnitAutomation.Steps
+{
+
+    public class LanguageSteps
+    {
+        string actualMessage;
+        string expectedMessage;
+        string languageName;
+        LanguageAddAndUpdateComponent languageAddAndUpdateComponent;
+        LanguageRenderingComponent languageRenderingComponent;
+        LanguageAssertHelper languageAssertHelper;
+        int cancelFlag;
+
+        public LanguageSteps()
+        {
+            actualMessage = string.Empty;
+            expectedMessage = string.Empty;
+            languageName = string.Empty;
+            cancelFlag = 0;
+            languageAddAndUpdateComponent = new LanguageAddAndUpdateComponent();
+            languageAssertHelper = new LanguageAssertHelper();
+            languageRenderingComponent = new LanguageRenderingComponent();
+
+        }
+        public void DeleteLanguageRecords()
+        {
+            languageAddAndUpdateComponent.DeleteAllLanguageRecords();
+            int rowCount = languageRenderingComponent.GetLanguageRecordsCount();
+            languageAssertHelper.AssertDeleteAllLanguageRecords(rowCount);
+        }
+        public void AddNewLanguage(LanguageDM languageDM)
+        {
+            languageAddAndUpdateComponent.DeleteAllLanguageRecords();
+            languageAddAndUpdateComponent.AddLanguage(languageDM);
+            actualMessage = languageRenderingComponent.CapturePopupMessage();
+            languageName = languageRenderingComponent.GetLastLanguageName();
+
+            if (!string.IsNullOrEmpty(languageName))
+                expectedMessage = languageName + " has been added to your languages";
+            else
+                expectedMessage = "has been added to your languages";
+            languageAssertHelper.AssertAddNewLanguage(expectedMessage, actualMessage);
+
+        }
+
+        public void AddNewLanguageRecordWithInsufficientData(LanguageDM languageDM)
+        {
+            languageAddAndUpdateComponent.CheckAndMakeSpaceAvailablity();
+            languageAddAndUpdateComponent.AddLanguage(languageDM);
+            actualMessage = languageRenderingComponent.CapturePopupMessage();
+            expectedMessage = "Please enter language and level";
+            languageAssertHelper.AssertNewLanguageNotAddedSuccessfully(expectedMessage, actualMessage);
+        }
+
+        public void AddAlreadyExistingLanguageRecord(LanguageDM languageDM)
+        {
+            languageAddAndUpdateComponent.DeleteAllLanguageRecords();
```

Comment:
so its nice your thinking about state but now i also want you change this so that it deletes the correct test data

example before adding language you need to make sure there a no existing languages as this could result in false positive test pass
you will need to think about this for all the scenarios

the state is say that you have a test that is adding a language english, the second time you run this test you should delete any existing language so that the state is reset, this will guarantee that your test does not produce false negatives and passes always
you can assume that test execution order is not always guaranteed
imagine if you run the tests  in parallel mode the all tests will run at the same time
so you need to also make sure you do not wipe out other tests data
hence why when you remove data the best thing to do is delete only data you added for the test to be correct
e.g test 1 added english and french as language test 2 added hindi then when you reset data you already know what language you are adding for each test
there for on reset of the two test you should delete english and french for test 1 and hindi for test 2

So to sum up you should delete all data if there is any before the start of any test and then after each test you should delete the data that a particular test is adding!
This is ensure that upon staring the test pack you have reset state and upon finishing each test deletes their data that they added!